### ISSUE TITLE
Tracking accessibility issues upstream

### ISSUE THREAD

**Issue by Malvoz** [Maps4HTML/Web-Map-Custom-Element#247]
Using code from Leaflet (and plugins) means this map component is sometimes inheriting accessibility issues.

We do have comments inline in code where fixes have been made, sometimes with links for upstream issues. Collecting them here in one spot helps get an overview of the state of accessibility in this component. This issue is to track upstream issues (some TBD) for discussion/fixes.

## Map component
- [x] Should be a programmatically identifiable element. Fixed in https://github.com/Maps4HTML/Web-Map-Custom-Element/pull/151/commits/7c86ddcbb8b39328b90930ffa1c5992f891f3df7. Upstream: https://github.com/Leaflet/Leaflet/issues/7193.
- [x] Use legible font-sizes. Fixed in https://github.com/Maps4HTML/Web-Map-Custom-Element/pull/210/commits/f7d705fe793c14fb409d07a3b4d020587af0f709. Upstream: https://github.com/Leaflet/Leaflet/pull/7346.
- [ ] The controls container should come before other author provided content such as popups in the DOM (i.e. `.leaflet-control-container` should come before `.leaflet-map-pane`). This reduces confusion by letting users form a consistent mental model of the content, across different maps (perhaps created by different authors, with different content). Relates to SC [1.3.2 Meaningful Sequence](https://www.w3.org/TR/WCAG21/#meaningful-sequence), [2.4.3 Focus Order](https://www.w3.org/TR/WCAG21/#focus-order) and [3.2.3 Consistent Navigation](https://www.w3.org/TR/WCAG21/#consistent-navigation). Upstream Leaflet: TBD.

## Controls
- [ ] `<a role="button">` should behave as `<button>`. Maybe re-open: https://github.com/Maps4HTML/Web-Map-Custom-Element/issues/19? Upstream: https://github.com/Leaflet/Leaflet/issues/7368 + https://github.com/Leaflet/Leaflet/issues/7369.
- [ ] Fullscreen control should be conveyed as a button, not a link. Upstream: https://github.com/runette/Leaflet.fullscreen/pull/2/ + https://github.com/Leaflet/Leaflet.fullscreen/pull/114.
- [ ] Popup's close button is missing an accessible name. Upstream: https://github.com/Leaflet/Leaflet/issues/7399.
- [x] Don't hide focus outline on keyboard navigation. Fixed in [`09db7d6`](https://github.com/Maps4HTML/Web-Map-Custom-Element/pull/108/commits/09db7d6dc0d544b9e0056d853b620d5a54662c4a#diff-56ad14054d43b64aa4b9c722e112e44b8769885fad4e9fc9311b63fea37cf643R73-R80). Upstream: https://github.com/Leaflet/Leaflet/pull/7259.
- [x] Unset the outline that Leaflet occasionally sets on controls without also delegating focus. Fixed in [`193cff0`](https://github.com/Maps4HTML/Web-Map-Custom-Element/pull/164/commits/193cff039da65a5efa17543f48221f8379c72550#diff-56ad14054d43b64aa4b9c722e112e44b8769885fad4e9fc9311b63fea37cf643R252-R256). Upstream: TBD.

## Context menu
- [x] Convey context menu items as buttons, not links. Fixed in [`5d076fe`](https://github.com/Maps4HTML/Web-Map-Custom-Element/commit/5d076fe4c827435bc54f86c9781c5207ce01f764#diff-70e49d69fe4e72eaeaf223302e53d9646e46d39ec4d02b8f31a6daba8bd0f92eR1285). Upstream: https://github.com/aratcliffe/Leaflet.contextmenu/issues/136.
- [ ] Delegate focus when context menu has been opened. Partial fix (only works in Firefox I believe) in [`95ecbcb`](https://github.com/Maps4HTML/Web-Map-Custom-Element/commit/95ecbcb2d2788194b380264dc2ef96a2839e5e9d#diff-70e49d69fe4e72eaeaf223302e53d9646e46d39ec4d02b8f31a6daba8bd0f92eR1375-R1377). Upstream: https://github.com/aratcliffe/Leaflet.contextmenu/issues/135.
- [ ] Delegate focus when context menu has been closed. Upstream: https://github.com/aratcliffe/Leaflet.contextmenu/issues/137.

**Comment by prushforth** [comment]
Regarding:

> Should be a programmatically identifiable element. Fixed in 7c86ddc. Upstream: Leaflet/Leaflet#7193.

Shouldn't this be on the mapml-viewer / map element?  The leaflet-container is in shadow DOM.

**Comment by Malvoz** [comment]
> Shouldn't this be on the mapml-viewer / map element?

The leaflet container has `tabindex=0` so giving that element the accessible name means it can be announced to the user as it is sequentially focused to.

**Comment by prushforth** [comment]
I'm just surprised that shadow DOM is traversable by AT, but probably because I am not too familiar with how it works and how it fails.

**Comment by Malvoz** [comment]
@prushforth if it wasn't then the content wouldn't be accessible to screen readers. I know that the ChromeVox screen reader (which is deprecated) has a bug where it doesn't read the content if `shadowRoot.mode = 'closed'`, rendering MapKit JS maps completely inaccessible to ATs.

I've updated the list btw.
